runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -623,26 +739,46 @@ function LanguageTree:set_included_regions(new_regions)
         region[i] = { range:range(true) }
       end
     end
-  end
+    ---@cast region Range6[]
+
+    -- Lookup the region.
+    local bucket = region_indices[region[1][3]]
+    local i ---@type integer?
+    if bucket then
+      for e = 1, #bucket, 2 do
+        if vim.deep_equal(bucket[e], region) then
+          i = bucket[e + 1] --[[@as integer]]
+          break
+        end
+      end
+    end
 
-  -- included_regions is not guaranteed to be list-like, but this is still sound, i.e. if
-  -- new_regions is different from included_regions, then outdated regions in included_regions are
-  -- invalidated. For example, if included_regions = new_regions ++ hole ++ outdated_regions, then
-  -- outdated_regions is invalidated by _iter_regions in else branch.
-  if #self:included_regions() ~= #new_regions then
-    -- TODO(lewis6991): inefficient; invalidate trees incrementally
-    for _, t in pairs(self._trees) do
-      self:_do_callback('changedtree', t:included_ranges(true), t)
+    -- If the region seems to be new, add it.
+    if not i then
+      i = #self._regions + 1 -- this always gives an unoccupied index even there are holes
+      if self._valid == true then
+        self._valid = {}
+        for j, _ in pairs(self._regions) do
+          self._valid[j] = true
+        end
+        self._valid[i] = false
+      elseif type(self._valid) == 'table' then
+        self._valid[i] = false
+      end
+      self._regions[i] = region
     end
-    self._trees = {}
-    self:invalidate()
-  else
-    self:_iter_regions(function(i, region)
-      return vim.deep_equal(new_regions[i], region)
-    end)
+
+    touched[i] = true
   end
 
-  self._regions = new_regions
+  -- Discard stale regions. If redrawing, defer it to _finish_batch.
```

Comment:
It first adds new regions and then discards stale regions. So the behavior of region indices are different from the old version. Some test failures are related to this. Sticking to the old behavior would add some overhead and sometimes just impossible due to batched cleanup.

Reply:
modified the tests to not assume how indices behave